lib/Serialization/Serialization.cpp

```diff
@@ -3688,11 +3688,17 @@ class Serializer::DeclSerializer : public DeclVisitor<DeclSerializer> {
     InlinableBodyTextLayout::emitRecord(S.Out, S.ScratchRecord, abbrCode, body);
   }
 
-  unsigned getNumberOfRequiredVTableEntries(
+  static bool getNeedsNewTableEntry(const AbstractFunctionDecl *func) {
+    if (isa_and_nonnull<ProtocolDecl>(func->getDeclContext()))
+      return func->requiresNewWitnessTableEntry();
```

Comment:
Calling `requiresNewWitnessTableEntry()` here is actually a change from the previous logic. Previously, if a protocol requirement used `override`, we would treat that as requiring a new witness table slot; now, we don't. Is the old behavior more correct than the new behavior? I'm not entirely clear on the behavior of `override` in a protocol, although I know we sometimes use it on associated types.

Reply:
From serialization's point of view, whether a declaration "requires a new vtable entry" just determines if we add a MissingMemberDecl in its place if the original decl won't deserialize. I'm not sure how this works in protocols though, it might not be wired up at all.

Explicit `override` on protocol members doesn't do anything by default, but we do compute overridden decls for protocol members, and the overriding decl (whether it has `override` on it or not!) is not part of the witness table's layout. There's a `-warn-implicit-overrides` to enable warnings if a protocol member is determined to be an override but isn't marked as `override`, and the stdlib builds with this flag, so that's why the stdlib makes use of `override` on protocol members.

In the requirement machine, all associated types with the same name are always part of the same equivalence class, so re-declaration has no effect. (We don't even query getOverridenDecl() on associated types, AFAIK.) However associated type inference will sometimes fail if you don't re-state the associated type -- that's a bug in associated type inference.

Value requirements determined to be overrides (again, whether they have the keyword or not) are also silently dropped. If you look at -emit-ir, you'll see that the witness table for Derived doesn't have entries for either f() and g(), and the calls inside `caller()` go through the `Base` conformance:

```
public protocol Base {
  func f()
  func g()
}

public protocol Derived: Base {
  func f()
  override func g()
}

public struct S: Derived {
  public func f() {}
  public func g() {}
}

func caller<T: Derived>(t: T) {
  t.f()
  t.g()
}
```